crypto/evp/evp_rand.c

```diff
@@ -338,32 +339,43 @@ EVP_RAND *EVP_RAND_CTX_rand(EVP_RAND_CTX *ctx)
     return ctx->meth;
 }
 
+static int evp_rand_get_ctx_params_locked(EVP_RAND_CTX *ctx,
+                                          OSSL_PARAM params[])
+{
+    return ctx->meth->get_ctx_params(ctx->data, params);
+}
+
 int EVP_RAND_get_ctx_params(EVP_RAND_CTX *ctx, OSSL_PARAM params[])
 {
-    int res = 1;
+    int res;
 
-    if (ctx->meth->get_ctx_params != NULL) {
-        if (!evp_rand_lock(ctx))
-            return 0;
-        res = ctx->meth->get_ctx_params(ctx->data, params);
-        evp_rand_unlock(ctx);
-    }
+    if (!evp_rand_lock(ctx))
+        return 0;
+    res = evp_rand_get_ctx_params_locked(ctx, params);
+    evp_rand_unlock(ctx);
     return res;
 }
 
+static int evp_rand_set_ctx_params_locked(EVP_RAND_CTX *ctx,
+                                          const OSSL_PARAM params[])
+{
+    /* Clear out the cache state because the values can change on a set */
+    ctx->strength = 0;
+    ctx->max_request = 0;
```

Comment:
Should this only happen if `set_ctx_params` is non-NULL?

Reply:
Only if `get_ctx_params` is non-NULL which is guaranteed by the changes you noted later.

Reply:
These two variables simply cache the values in the provider rand.
It's tempting to remove them and call through all the time.

Reply:
I've gone ahead and removed these two.  They are now always queried from the underlying PROV_RAND object.
This will avoid any possible cache incoherency problems.

Reply:
I don't quite get the reason for removing the caching. Originally, it was a speed optimization, I guess. Now it has been removed on the ground taht it avoids cache incoherency, even though `evp_rand_set_ctx_params_locked` is holding the rand lock (if enabled).